### PR TITLE
Features/vector search rag pipeline

### LINKED ISSUE #29: Add pipeline using vector search for RAG
Even an unoptimised vector search performs pretty well, and using results for Retrieval-Augmented Generation will make it more useful

### PR DESCRIPTION
**Pull request by kuraisle**
## Pull Request Contents

| |
|-|
♻️ Refactor
✨ Feature
📝 Documentation

## Description

This initial version of the pipeline was only for testing, but adds some useful functionality

- Added the endpoint to the app.py
- Added methods to the embeddings class so that matches can be retrieved from the vector database
- Modified the models.py so that its return type is a Union of the GPT and Llama.cpp generators
- Added the `get_RAG_assistant` method to the pipeline.py. Consistent with the existing `get_basic_assistant` method
- Modified the prompts to be compatible with llama 3.1. Be warned, this prompt format doesn't work with previous llamas. To be fixed in another PR

## Related Issues or other material

- [X] This PR relates to one or more issues, detailed below

- Related Issue #29 
- Closes #29 

## ✅ Added/updated tests?

- [X] This PR contains relevant tests
  - Or doesn't need to per the below explanation

Still no tests

- [X] I've completed all **actions** and **tasks** detailed in the PR Template

### CODE REVIEW

File: Carrot-Assistant/components/models.py [hunk 1]
```diff
@@ -27,7 +28,7 @@
 
 def get_model(
     model_name: str, temperature: float = 0.7, logger: logging.Logger | None = None
-) -> object:
+) -> OpenAIGenerator | LlamaCppGenerator:
```
[the remark this answers is not in the document]
Comment (kuraisle) [reply Carrot-Assistant/components/models.py:31]: Nope, see PEP 604

File: Carrot-Assistant/app.py [hunk 2]
```diff
@@ -92,7 +93,7 @@ class PipelineOptions(BaseModel):
     force_rebuild: bool = False
     embed_vocab: list[str] = ["RxNorm", "RxNorm Extension"]
```
[the remark this answers is not in the document]
Comment (kuraisle) [reply Carrot-Assistant/app.py]: Done

**Review by CodeByKarthik — APPROVED** [review]
Hi James, 

I have no issues with the code. I'll merge it on the main. Good Job👍🏻

**Comment by Karthi-DStech** [comment]
Hi James, 

I can see there are some conflicts in the branch. Could you please solve it?

**Pull request merged**